lambda-buffers-proto/compiler.proto

```diff
@@ -633,6 +633,75 @@ message ProtoParseError {
   }
 }
 
+// Non-internal errors that can be thrown during
+// type class checking. Because these errors involve
+// TyClassCheck types, which cannot be converted back into
+// proto messages, the error message is constructed in the
+// TypeClassCheck modules in LambdaBuffers.Compiler
+message TypeClassError {
+  // Error which indicates that a class referred to in an instance
+  // or deriving clause does not exist. (Maybe this should be internal?)
+  message UnknownClassError {
+    string msg = 1;
+    SourceInfo source_info = 2;
+  }
+
+  // Error which indicates that the TyDef referred to by a TyRef cannot be found
+  message LocalTyRefNotFoundError {
+    string msg = 1;
+    // SourceInfo points to the clause that triggered the tyref lookup
+    SourceInfo source_info = 2;
+  }
+
+  // Error which indicates a cycle of superclasses. SourceInfo
+  // isn't particularly meaningful here - we could include the SourceInfo
+  // for each class in the cycle, but given the large amount of work
+  // required to define a LB class, this should be more than adequate
+  message SuperclassCycleError {
+    string msg  = 1;
+  }
+
+  // Error that indicates a constraint cannot be solved with
+  // the in-scope set of instance rules.
+  message FailedToSolveConstraintError {
+    string msg = 1;
+    // SourceInfo points to the clause that triggered the subgoal failure
+    // (The failure could potentially be a structural rule, so pointing
+    // to the clause that triggered the exception is the best we can do)
+    SourceInfo source_info = 2;
+  }
```

Comment:
I think this is the key error right? This is what we need if so:

```
// When solving a  `constraint`, a type class instance was found to be missing to solve the `missing_constraint`  in `module_name`.
message MissingInstanceError {
  ModuleName module_name = 1;
  Constraint constraint = 2;
  Constraint missing_constraint = 2;
}
```

Reply:
```
Constraint missing_constraint = 2;
```

This is the problem (I'm assuming the first field 2 is the constraint that we're trying to solve and that the second is the actual failure case). 

The second field has a SourceInfo always because it's constructed from a `P.InstanceClause` or `P.DeriveClause` or whatever. 

The third field will always end up being one of the `assumptions`  (from `Validate.hs`): 

```
    assumptions =
      S.toList . S.fromList $
        S.toList inScopeInstances -- *A* the basic set of in-scope instances
          <> concatMap (mkStructuralRules . constraintClass) (c' : cs') -- *B* structural rules for all in scope classes
          <> S.toList (S.filter (/= i) . S.mapMonotonic unTag $ localInstances) -- *C* all instances in the current module that aren't the one we're trying to check
```

A and C can be tagged with a `PC.Constraint` since they're constructed from that originally. *B* can't be (and it is possible to end up with a failure involving Nil/Cons lists and variables "in the wrong place" for a `PC.Ty`). 

I can only think of two ways around this: 

1. Express the structural rules operationally instead of declaratively, which is bad because then you're definitely not getting a printout of them like you wanted. 
2. Tag everything with a `Maybe PC.Constraint` and have a meta-error type for... malformed errors... which seems incredibly stupid.

Reply:
Yeah what you're suggesting here is actually impossible. We also check that superclass constraints are satisfied as part of constraint solving, and those superclass constraints are always implicit / do not ever exist in the source. `missing_constraint` cannot contain a proto `Constraint` and expect the sourceinfo to mean anything at all.